### PR TITLE
Title: Scheme/parsec-tpm: Add TPM vendor and model information

### LINKED ISSUE #139: Parsec TPM endorsements - consider extending with vendor and model info
@yogeshbdeshpande [comment](https://github.com/veraison/services/pull/134#discussion_r1165662506_):
```
              We should have some **optional**  Vendor or Model as text associated to class
to describe an Attester, in addition to UEID()?
```

### PR DESCRIPTION
**Pull request by Sukuna0007Abhi**
Description:
This PR implements support for TPM vendor and model information in TPM endorsements, addressing Issue #139.

Changes include:
- Extended TaAttr structure with vendor and model fields
- Added input validation and sanitization for security
- Implemented comprehensive test coverage:
  * Real-world TPM vendor/model combinations
  * International character handling
  * Security test cases
  * Edge cases and input validation
- Updated documentation with new fields and security considerations

Security Considerations:
- Implemented strict input validation
- Added sanitization for potentially dangerous inputs
- Protected against injection attacks
- Enforced character and length limits

Testing:
- Added extensive test suite covering:
  * Common TPM vendors (Infineon, STMicroelectronics, etc.)
  * International vendor names (Unicode support)
  * Security test cases (injection attempts)
  * Edge cases and combinations
  * Character type handling

Documentation:
- Updated README.md with new field descriptions
- Added security considerations
- Included usage examples

Resolves: #139 

Ready for Review sir @thomas-fossati @setrofim @yogeshbdeshpande

### CODE REVIEW

File: scheme/parsec-tpm/README.md [hunk 1]
```diff
@@ -24,7 +24,116 @@
...
+          "instance-id": {
+            "type": "ueid",
+            "value": "AQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
+          }
+        }
+      },
+      "key": [{
+        "type": "pkix-base64-key",
+        "value": "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAETKRFE_RwSXooI8DdatPOYg_uiKm2XrtT_uEMEvqQZrwJHHcfw0c3WVzGoqL3Y_Q6xkHFfdUVqS2WWkPdKO03uw==",
+        "parameters": {
```
[the remark this answers is not in the document]
Comment (setrofim) [reply scheme/parsec-tpm/README.md:70]: `crypto-key-type-choice` offers "type choice" rather than "map" extension point. Meaning you cannot add fields to existing values, you can only add new types. So you cannot add parameters to, e.g., `"pkix-base64-key"`, you'd need to define a whole new type, `"pkix-base64-key-with-params"` and register that as an extension to the type choice (and then do the same for every other variant that needs to be extended in this way).

[the remark this answers is not in the document]
Comment (Sukuna0007Abhi) [reply scheme/parsec-tpm/README.md]: Ready for review sir

**Review by setrofim — APPROVED** [review]
LGTM

File: scheme/parsec-tpm/sanitize.go [hunk 2]
```diff
@@ -0,0 +1,213 @@
+// Copyright 2024 Contributors to the Veraison project.
```
Comment (yogeshbdeshpande) [inline scheme/parsec-tpm/sanitize.go:1]:
```suggestion
// Copyright 2025 Contributors to the Veraison project.
```

File: scheme/parsec-tpm/sanitize_test.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
+// Copyright 2024 Contributors to the Veraison project.
```
Comment (yogeshbdeshpande) [inline scheme/parsec-tpm/sanitize_test.go:1]:
```suggestion
// Copyright 2025 Contributors to the Veraison project.
```

File: scheme/parsec-tpm/sanitize.go (hunk 2 above)
Comment (yogeshbdeshpande) [inline scheme/parsec-tpm/sanitize.go:1]: I am wondering this generic implementation of string validation and checking for specific patterns is applicable more generically across entire Services `scheme` and could be moved to `services/common` ??

**Review by yogeshbdeshpande — COMMENTED** [review]
Please check my comments!

File: scheme/parsec-tpm/sanitize.go (hunk 2 above)
Comment (yogeshbdeshpande) [reply scheme/parsec-tpm/sanitize.go:1]: ok, understood, thanks. Are these then coming from a spec reference for restriction in TPM specific scene setting?

Comment (setrofim) [reply scheme/parsec-tpm/sanitize.go:1]: Not that I'm aware of (as I've said -- I'm not familiar with the specs). But this looks like it's designed specifically to catch "unreasonable" vendor/model names. From that point of view, the implementation looks reasonable. But I would not accept this as a good generic santizer (not do I think it's necesarily a good idea to have one).

**Comment by Sukuna0007Abhi** [comment]
Sure sir, @yogeshbdeshpande will try my best, to explain properly 🙏

File: scheme/parsec-tpm/README.md [hunk 4]
```diff
@@ -24,7 +24,166 @@
...
+
+The TPM Trust Anchor can include optional vendor and model information to provide additional context about the TPM manufacturer and specific model. These fields are:
+
+- `parsec-tpm.vendor`: The manufacturer or vendor of the TPM (optional)
+- `parsec-tpm.model`: The specific model identifier of the TPM (optional)
+
+Both fields support:
+- Unicode characters (e.g., for international vendor names)
+- Special characters (allowed in both fields)
+- Variable length strings (no artificial length restrictions)
```
Comment (yogeshbdeshpande) [inline scheme/parsec-tpm/README.md:44]:
```suggestion
Both fields support:
- Unicode characters (e.g., for international vendor names)
- Special characters (allowed in both fields)
- Variable length strings (no artificial length restrictions)
```

File: scheme/parsec-tpm/README.md [hunk 5]
```diff
@@ -24,7 +24,166 @@
...
+      },
+      "key": [{
+        "type": "pkix-base64-key",
+        "value": "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAETKRFE_RwSXooI8DdatPOYg_uiKm2XrtT_uEMEvqQZrwJHHcfw0c3WVzGoqL3Y_Q6xkHFfdUVqS2WWkPdKO03uw=="
+      }]
+    }
+  ]
+}
+```
+
```
Comment (yogeshbdeshpande) [inline scheme/parsec-tpm/README.md:77]: I do not think we need additional example, just drop them

File: scheme/parsec-tpm/corim_extractor.go [hunk 6]
```diff
@@ -121,6 +122,27 @@ func makeTaAttrs(id ID, key *comid.CryptoKey) (json.RawMessage, error) {
 		"parsec-tpm.ak-pub":      key.String(),
 	}
 
+	// Extract optional vendor and model from environment.class
+	// Following CoRIM specification - vendor/model are stored in environment, not key parameters
+	if env.Class != nil {
+		if env.Class.Vendor != nil {
+			vendor := string(*env.Class.Vendor)
+			// Trim and validate
+			vendor = sanitizeAndValidate(vendor)
```
Comment (yogeshbdeshpande) [inline scheme/parsec-tpm/corim_extractor.go:131]: Line 131 not needed!

File: scheme/parsec-tpm/corim_extractor.go [hunk 7]
```diff
@@ -121,6 +122,27 @@ func makeTaAttrs(id ID, key *comid.CryptoKey) (json.RawMessage, error) {
...
+			if vendor != "" {
+				attrs["parsec-tpm.vendor"] = vendor
+			}
+		}
+		if env.Class.Model != nil {
+			model := string(*env.Class.Model)
+			// Trim and validate
+			model = sanitizeAndValidate(model)
```
Comment (yogeshbdeshpande) [inline scheme/parsec-tpm/corim_extractor.go:139]: Line 139 not needed!

File: scheme/parsec-tpm/corim_extractor.go [hunk 8]
```diff
@@ -188,3 +210,22 @@ func (o *ID) FromEnvironment(e comid.Environment) error {
 func (o *CorimExtractor) SetProfile(profile string) {
 	o.Profile = profile
 }
+
```
Comment (yogeshbdeshpande) [inline scheme/parsec-tpm/corim_extractor.go:213]: Remove Lines 213 to 231

File: scheme/parsec-tpm/sanitize.go [hunk 9]
```diff
@@ -0,0 +1,213 @@
+// Copyright 2025 Contributors to the Veraison project.
```
Comment (yogeshbdeshpande) [inline scheme/parsec-tpm/sanitize.go:1]: File can be deleted!

**Review by yogeshbdeshpande — CHANGES_REQUESTED** [review]
See my all comments, please re-work on the PR!

**Comment by THS-on** [comment]
General no response from @Sukuna0007Abhi on other PRs, so closing.
